### PR TITLE
Add the deferrable mode to RunPipelineJobOperator

### PR DESCRIPTION
**Pull request by e-galan**
- Add the deferrable mode to RunPipelineJobOperator
- Add new tests, correct some of the existing tests

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/vertex_ai/pipeline_job.py [hunk 1]
```diff
@@ -178,11 +178,17 @@ def run_pipeline_job(
         service_account: str | None = None,
         network: str | None = None,
         create_request_timeout: float | None = None,
-        experiment: str | experiment_resources.Experiment | None = None,
```
Comment (MaksYermak) [inline airflow/providers/google/cloud/hooks/vertex_ai/pipeline_job.py:181]: What is the reason for removing this parameter? Also we can't remove parameters without deprecating them first.

Comment (e-galan) [reply airflow/providers/google/cloud/hooks/vertex_ai/pipeline_job.py:181]: This parameter is no longer used in any of the underlying methods. Is there any procedure for deprecation?

Comment (e-galan) [reply airflow/providers/google/cloud/hooks/vertex_ai/pipeline_job.py]: Parameter kept.

File: airflow/providers/google/cloud/triggers/vertex_ai.py [hunk 2]
```diff
@@ -89,11 +91,97 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
...
 
     def _get_async_hook(self) -> HyperparameterTuningJobAsyncHook:
         return HyperparameterTuningJobAsyncHook(
             gcp_conn_id=self.conn_id, impersonation_chain=self.impersonation_chain
         )
+
+
+class RunPipelineJobTrigger(BaseTrigger):
+    """A trigger that makes async calls to Vertex AI to check the state of a running pipeline job."""
+
```
Comment (e-galan) [inline airflow/providers/google/cloud/triggers/vertex_ai.py:106]: I am not sure about what states can be called successful here. I saw that in some other cases (one is CreateHyperparameterTuningJobTrigger) the Paused state is considered a successful one, but I don't understand why. Can somebody clear this up for me, please?

File: airflow/providers/google/cloud/triggers/vertex_ai.py [hunk 3]
```diff
@@ -89,11 +91,97 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
...
+    SUCCESSFUL_PIPELINE_STATES = (PipelineState.PIPELINE_STATE_SUCCEEDED,)
```
Comment (e-galan) [inline airflow/providers/google/cloud/triggers/vertex_ai.py:107]: I am not sure about what states can be called successful here. I saw that in some other cases (one is CreateHyperparameterTuningJobTrigger) the Paused state is considered a successful one, but I don't understand why. Can somebody clear this up for me, please?

Comment (moiseenkov) [reply airflow/providers/google/cloud/triggers/vertex_ai.py:107]: Although the status model might change in future, I'd rely on statuses presented in [SDK](https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform/jobs.py#L69C1-L69C21). They are grouped as "complete", "error", and "pending". So my suggestion is to consider successfull states as "complete" states excluding "error" states.

File: airflow/providers/google/cloud/hooks/vertex_ai/pipeline_job.py [hunk 4]
```diff
@@ -407,3 +503,86 @@ def delete_pipeline_job(
...
+
+    async def get_credentials(self) -> Credentials:
+        credentials = (await self.get_sync_hook()).get_credentials()
+        return credentials
+
+    async def get_project_id(self) -> str:
+        sync_hook = await self.get_sync_hook()
+        return sync_hook.project_id
+
+    async def get_project_location(self) -> str:
```
Comment (moiseenkov) [inline airflow/providers/google/cloud/hooks/vertex_ai/pipeline_job.py:537]:
```suggestion
    async def get_location(self) -> str:
```

File: airflow/providers/google/cloud/operators/vertex_ai/pipeline_job.py [hunk 5]
```diff
@@ -82,11 +87,13 @@ class RunPipelineJobOperator(GoogleCloudBaseOperator):
...
-    :param experiment: Optional. The Vertex AI experiment name or instance to associate to this
-        PipelineJob. Metrics produced by the PipelineJob as system.Metric Artifacts will be associated as
-        metrics to the current Experiment Run. Pipeline parameters will be associated as parameters to
+    :param experiment: Optional. The Vertex AI experiment name or instance to associate to this PipelineJob.
+        Metrics produced by the PipelineJob as system.Metric Artifacts will be associated as metrics
+        to the current Experiment Run. Pipeline parameters will be associated as parameters to
         the current Experiment Run.
     :param gcp_conn_id: The connection ID to use connecting to Google Cloud.
+    :param sync: Whether to execute this method synchronously. If False, this method will unblock, and it
+        will be executed in a concurrent Future. The default is True.
```
Comment (moiseenkov) [inline airflow/providers/google/cloud/operators/vertex_ai/pipeline_job.py:96]: Haven't we agreed to remove this parameter?

**Pull request closed without merging**